common/persistence/nosql/nosqlplugin/cassandra/translator/fixed_address_translator.go

```diff
@@ -0,0 +1,84 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package translator
+
+import (
+	"errors"
+	"fmt"
+	"net"
+
+	"github.com/gocql/gocql"
+	"go.temporal.io/server/common/config"
+)
+
+const (
+	fixedTranslatorName   = "fixed-address-translator"
+	advertisedHostnameKey = "advertised-hostname"
+)
+
+func init() {
+	RegisterTranslator(fixedTranslatorName, NewFixedAddressTranslatorPlugin())
+}
+
+type FixedAddressTranslatorPlugin struct {
+}
+
+func NewFixedAddressTranslatorPlugin() TranslatorPlugin {
+	return &FixedAddressTranslatorPlugin{}
+}
+
+func (plugin *FixedAddressTranslatorPlugin) GetTranslator(cfg *config.Cassandra) (gocql.AddressTranslator, error) {
```

Comment:
Could you paste what you explained to me in slack message as comments to here.

Reply:
_I might be missing something, but the implementation of FixedAddressTranslator just lookup IP address from an advertised-hostname, which is from a static config. So the IP address used by gocql is just that IP from that configured advertised-hostname._

this is correct

_Just for me to understand this better, can you just config your persistence to directly use advertised-hostname as your server address?_

As I said in my last comment, it seems to me that the service resolver is doing this thing only, https://github.com/temporalio/temporal/blob/master/common/persistence/nosql/nosqlplugin/cassandra/gocql/client.go#L50

This is used for cases when you have "hostname1" in the configuration but that node is actually running at "hostname2", so you translate it. This translation is used only for cases when you need to translate hostnames of contact points (aka "seeds" in your terminology)

So then you take these translated hostnames and you create gocql-specific NewCluster, here https://github.com/temporalio/temporal/blob/master/common/persistence/nosql/nosqlplugin/cassandra/gocql/client.go#L52

What cql driver does is that it will connect to the first node in that list (if there is more than one), if it fails to connect to it, it will pick another from that list and so on so, when the connection is initialised, the driver does not know what the topology of the cluster looks like yet - it just connected to a node.

In order to know, for the driver itself, internally, what the topology is like, it will query that node it just connected to and it will read system.peers table. In that table, there are all other nodes of a cluster as that node you just connected to sees it. Every other node has the very same system.peers table where all other nodes of the cluster are. The returned rows, representing all other nodes in the cluster, contain IP addresses internal of that cluster. They are not necessarily the same hostnames as the ones you would translate with your service resolver (they are IP addresses, not hostnames, actually).

For the case of the private link, or if you have them behind the proxy, service resolver would translate just the publicly visible hostname which you put into configuration so driver can connect to it, but that is not enough, ip addresses behind a proxy are not reachable from client's perspective, they are just that, internal ... these ones are not translatable with service resolver so client cant connect to them directly. 

That is what gocql address translator is exactly for https://github.com/temporalio/temporal/blob/9fd57308c23ce74e51e872f7ab67643e7ae1ad[…]/common/persistence/nosql/nosqlplugin/cassandra/gocql/client.go

cluster.AddressTranslator is gocql specific thingy. You had good intentions to provide service resolver for Cassandra persistence to translate the addresses, it is just "not enough". When it comes to address translation gocql offers itself for internal purposes, your service resolver does not cover that case, it translates only your "seeds".

Reply:
@smiklosovic I should be more clear, I mean to put those content (extract the relevant part) in code as comments. It has valuable info that helps reader to understand why this is needed.

Reply:
@yiminc  added, tell me if you want to somehow chop it but I think for overall context all parts are important.